source-code/core/src/lint/linter.ts

```diff
@@ -0,0 +1,193 @@
+import type { Resource, Message, Pattern } from '../ast/schema.js'
+import type { Config, EnvironmentFunctions } from '../config/schema.js'
+import { createContext } from './context.js';
+import { getLintRulesFromConfig, LintableNode, LintableNodeByType, ConfiguredLintRule, NodeVisitor, NodeVisitors, TargetReferenceParameterTuple } from './rule.js';
+
+const getResourceForLanguage = (resources: Resource[], language: string) =>
+	resources.find(({ languageTag }) => languageTag.name === language);
+
+export const lint = async (config: Config, env: EnvironmentFunctions) => {
+	const { referenceLanguage, languages, readResources } = config
+
+	const lintRules = getLintRulesFromConfig(config)
```

Comment:
Using one liner functions like `getLintRulesFromConfig` makes the code harder to read and bloats the implementation IMHO. 

The following is much easier to reason about:

```ts
const lintRules = config.lint?.rules
// vs 
const lintRules = getLintRulesFromConfig(config)
```

Even if the array is flattened:

```ts
const lintRules = config.lint?.rules ? config.lint.rules.flat() : []
// or
const lintRules = (config.lint?.rules ?? []).flat()

// vs 
const lintRules = getLintRulesFromConfig(config)
```

The conditional logic of whether or not rules are undefined and the `lintRules` variable are not even required if the if statement is moved to the top of the function: 

```ts
export const lint = async (config: Config, env: EnvironmentFunction) => {
    if (config.lint?.rules === undefined) {
		console.warn('No lint rules specified. Aborting ...')
		return
	}
    // ...
    await Promise.all(
        // use config rules directly
		config.lint.rules.flat().map(lintRule =>
			processLintRule({ env, lintRule, referenceLanguage, languages, reference, resources })
		)
	)
}

Reply:
If someone would have this config:
```ts
{
	lint: []
}
```
then your code will not print a warning.

I could inline the `getLintRulesFromConfig` function, but if a line of code does multiple things (check if `lint` property is present, return the rules, flatten them), it should be it's own function with a meaningful name.

Reply:
Adding a conditional check to the top level `if` seems like a potential fix?  

```ts
export const lint = async (config: Config, env: EnvironmentFunction) => {
    if (config.lint?.rules === undefined || config.lint.rules.length === 0) {
		console.warn('No lint rules specified. Aborting ...')
		return
	}
    const rules = config.rules.flat()
    // ...
}
```

Reply:
Regardless of this specific example, I found the code harder to read than it should have but that might just be me. 

Hiding one line of code behind a function that is only used once leads to: 

- harder to reason about.  the reader needs to jump into the function instead of just "seeing" what the line of code does.
- more code. the function needs to be defined, imported, called. 
- (the calling function will not be pure, but this is minor)

Reply:
> ```ts
> if (config.lint?.rules === undefined || config.lint.rules.length === 0) {
> ```

And I find this harder to read and reason about than a function with the name `getLintRulesFromConfig` and a `length` check afterwards ^^.

> * harder to reason about.  the reader needs to jump into the function instead of just "seeing" what the line of code does.

The question is: does someone really have to know how lint rules are specified in the config? If I see some code that says `readInlangConfig`, I trust that function to do what it tells and I don't care about the implementation. With the beauty of `TypeScript` I can immediately see if the function returns what I would expect without ever seeng the code of that function.

> * more code. the function needs to be defined, imported, called.

The code will probably be minimized, and will end up with less characters than your suggested alternative above. Even without minification, the function approach takes an equal amount of characters. And a single function call will not make the code noticeable slower ;)

The benefit of doing this in it's own function is, that we can test it in isolation and if the config changes, we only need to update a single function.

If the name of the function is unclear, feel free to suggest an alternative.

Reply:
You are implementing this, thus go ahead and do what you find better.

This seems like a preference. I find a one-line function more difficult to read for something so trivial as a conditional. I'd rather see the conditional as a developer and know exactly what is going on.  

> does someone really have to know how lint rules are specified in the config? 

Yes, please. Makes it obvious where the rules are coming from. One core concept of UX is discoverability. Does a user understand how something works? 

Imagine that I am looking at this code on GitHub or similar and wonder where the rules are coming from. I am out of luck (if GitHubs new reference feature is not working) and have to do extra work to understand where rules are coming from. 

>> more code. the function needs to be defined, imported, called.
> The code will probably be minimized, and will end up with less characters than your suggested alternative above.

I was referring to code to maintain and comprehend, not the production build.

Reply:
As discussed privately, I moved the function to the same file to improve the point of "discoverability".